OpenSim/Simulation/Wrap/PathWrapPoint.cpp

```diff
@@ -0,0 +1,124 @@
+
+OpenSim::PathWrapPoint::PathWrapPoint() = default;
+
+void OpenSim::PathWrapPoint::extendAddToSystem(SimTK::MultibodySystem& system) const
+{
+    Super::extendAddToSystem(system);
+
+    _wrapPath = addCacheVariable("wrap_path", Array<SimTK::Vec3>{}, SimTK::Stage::Position);
+    _wrapPathLength = addCacheVariable("wrap_path_length", 0.0, SimTK::Stage::Position);
+    _location = addCacheVariable("wrap_location", SimTK::Vec3{}, SimTK::Stage::Position);
+}
+
+const OpenSim::Array<SimTK::Vec3>& OpenSim::PathWrapPoint::getWrapPath(const SimTK::State& s) const
+{
+    return getCacheVariableValue(s, _wrapPath);
+}
+
+void OpenSim::PathWrapPoint::setWrapPath(const SimTK::State& s, const Array<SimTK::Vec3>& src) const
+{
+    updCacheVariableValue(s, _wrapPath) = src;
+    markCacheVariableValid(s, _wrapPath);
+}
+
+void OpenSim::PathWrapPoint::clearWrapPath(const SimTK::State& s) const
+{
+    updCacheVariableValue(s, _wrapPath).setSize(0);
+    markCacheVariableValid(s, _wrapPath);
+}
+
+double OpenSim::PathWrapPoint::getWrapLength(const SimTK::State& s) const
+{
+    return getCacheVariableValue(s, _wrapPathLength);
+}
+
+void OpenSim::PathWrapPoint::setWrapLength(const SimTK::State& s, double aLength) const
+{
+    updCacheVariableValue(s, _wrapPathLength) = aLength;
+    markCacheVariableValid(s, _wrapPathLength);
+}
+
+const OpenSim::WrapObject* OpenSim::PathWrapPoint::getWrapObject() const
+{
+    return _wrapObject.get();
+}
+
+void OpenSim::PathWrapPoint::setWrapObject(const WrapObject* wrapObject)
+{
+    _wrapObject.reset(wrapObject);
+}
+
+SimTK::Vec3 OpenSim::PathWrapPoint::getLocation(const SimTK::State& s) const
+{
+    return getCacheVariableValue(s, _location);
+}
+
+void OpenSim::PathWrapPoint::setLocation(const SimTK::State& s, const SimTK::Vec3& loc) const
+{
+    updCacheVariableValue(s, _location) = loc;
+    markCacheVariableValid(s, _location);
+}
+
+SimTK::Vec3 OpenSim::PathWrapPoint::getdPointdQ(const SimTK::State& s) const
+{
+    return SimTK::Vec3(0);
+}
+
+SimTK::Vec3 OpenSim::PathWrapPoint::calcLocationInGround(const SimTK::State& state) const
+{
+    return getParentFrame().getTransformInGround(state)*getLocation(state);
+}
+
+SimTK::Vec3 OpenSim::PathWrapPoint::calcVelocityInGround(const SimTK::State& state) const
+{
+    // compute the local position vector of the station in its reference frame
+    // expressed in ground
+    SimTK::Vec3 r = getParentFrame().getTransformInGround(state).R()*getLocation(state);
+    const SimTK::SpatialVec& V_GF = getParentFrame().getVelocityInGround(state);
+
+    // The velocity of the station in ground is a function of its frame's
+    // linear (vF = V_GF[1]) and angular (omegaF = A_GF[0]) velocity, such that
+    // velocity of the station: v = vF + omegaF x r
+    return V_GF[1] + V_GF[0] % r;
+}
+
+SimTK::Vec3 OpenSim::PathWrapPoint::calcAccelerationInGround(const SimTK::State& state) const
+{
+    // The spatial velocity of the reference frame expressed in ground
```

Comment:
These codeblocks were ripped from `OpenSim::Station`, but they don't appear to be ideally implemented.

The existing implementation repeatably looks up the parent frame, which might involve runtime lookups. I also need to check how frames typically store their data, because this entire algorithm is putting things in ground one-at-a-time, whereas it might be more efficient to:

- Grab a reference to the parent frame
- Create a stack-local `transformInGround`
- Use the transform to transform the (not-in-ground) velocity, acceleration, and location

This could be faster because it only involves one frame lookup, one transform creation, and several trivial getter calls (for the "raw" velocity, acceleration, etc.)

Reply:
Dumping fixing this in this PR because it has a negligible perf change